moit-domain/src/main/kotlin/com/mashup/moit/domain/study/StudyDatesCalculator.kt

```diff
@@ -0,0 +1,45 @@
+package com.mashup.moit.domain.study
+
+import com.mashup.moit.domain.moit.ScheduleRepeatCycle
+import org.springframework.stereotype.Component
+import java.time.DayOfWeek
+import java.time.LocalDate
+import java.time.LocalDateTime
+import java.time.LocalTime
+
+@Component
+class StudyDatesCalculator {
+    fun calculateStudyDates(
+        dayOfWeeks: Set<DayOfWeek>,
+        startDate: LocalDate,
+        endDate: LocalDate,
+        startTime: LocalTime,
+        repeatCycle: ScheduleRepeatCycle,
+    ): List<LocalDate> {
+        val firstWeekDates = calculateFirstWeekStudyDates(
+            dayOfWeeks = dayOfWeeks,
+            startDate = startDate,
+            startTime = startTime,
+        )
+
+        if (repeatCycle.period == null) return firstWeekDates.filter { !it.isAfter(endDate) }
+
+        return generateSequence(firstWeekDates) { dates -> dates.map { it.plus(repeatCycle.period) } }
+            .takeWhile { dates -> dates.any { !it.isAfter(endDate) } }
+            .flatten()
+            .filter { !it.isAfter(endDate) }.toList()
+    }
+
+    private fun calculateFirstWeekStudyDates(
+        dayOfWeeks: Set<DayOfWeek>,
+        startDate: LocalDate,
+        startTime: LocalTime,
+    ): List<LocalDate> {
+        return generateSequence(startDate) { date -> date.plusDays(1) }
+            .filter { date ->
+                date.dayOfWeek in dayOfWeeks
+                        && LocalDateTime.of(date, startTime).isAfter(LocalDateTime.now())
+            }
+            .take(dayOfWeeks.size).toList()
+    }
```

Comment:
`it` 중첩되는 것 없으면 바로 사용해도 될 듯?
```suggestion
        return generateSequence(startDate) { it.plusDays(1) }
            .filter {
                it.dayOfWeek in dayOfWeeks 
                        && LocalDateTime.of(it, startTime).isAfter(LocalDateTime.now())
            }
            .take(dayOfWeeks.size).toList()
```

Reply:
혹시 그리고 이거 로직 설명 좀 해줄 수 있어? 목적이 뭐지...?